File: Leetcode/JS/Leetcode851-喧闹和富有.js
var loudAndRich = function(richer, quiet) {
    let n = quiet.length;
    let ans = new Array(n).fill(-1);
    let g = new Array(n).fill(0).map(() => []);
    for(const rich of richer) {
        let a = rich[0],b = rich[1];
        g[b].push(a);
    }
    let dfs = function(u) {
        if(ans[u] != -1) return
        ans[u] = u;
        for(const v of g[u]){
            dfs(v);
            if(quiet[ans[v]] < quiet[ans[u]]){
                ans[u] = ans[v];
            }
        }
    }
    for(let i = 0;i < n;i++) dfs(i);
    return ans;
}

File: Leetcode/PY/Leetcode851-喧闹和富有.py
class Solution:
    def loudAndRich(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        n = len(quiet)
        ans = [-1] * n
        g = [[] for _ in range(n)]

        def dfs(u):
            if ans[u] != -1:return
            ans[u] = u
            for i in g[u]:
                dfs(i)
                if quiet[ans[i]] < quiet[ans[u]]:
                    ans[u] = ans[i]

        for rich in richer:
            a,b = rich[0],rich[1]
            g[b].append(a)
        for i in range(n):
            dfs(i)
        return ans